ls1intum/Artemis
# `Plagiarism checks`: Introduce continuos plagiarism control

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
Continuous plagiarism control (cpc) introduces possibility of a new approach for preventing plagiarism in student submissions. Cpc automatically scans all submissions and gently informs students, whose solutions are very similar to solutions of any other student.

Currently students receive information about a plagiarism case only after the exercise due date. This information itself has a form of punishment, as it leaves no room for improvement and can have severe consequences for the student.  The new approach shortens the feedback loop of plagiarism checks and therefore encourages students to improve their solutions before the submission time is over. This may lead to less plagiarism cases opened for each exercises which improves the overall learning outcomes for students and reduces the amount of work to be performed by instructors.

### Description

This is an experimental feature.

Continuous plagiarism control consist of the following features:
- plagiarism detection can start automatically each night during the submission period,
- optionally one more detection starts the night after the submission period is over,
- Artemis creates plagiarism cases for submissions included in plagiarism comparisons detected by the cpc,
- cpc plagiarism cases posts and notifications contain special messages that only mention possible similarities to other submissions (see screenshots for details)
- Artemis removes cpc plagiarism cases from submissions when any the subsequent cpc runs detects similarity is gone,
- Artemis stores configuration for cpc checks in the exercise model,
- instructors still can trigger manual plagiarism checks and configure them per run using the current interface.

> [!NOTE]  
> For those who reviewed other cpc related PRs in the past: This PR introduced new, simplified version of the cpc. The general approach is unchanged and it contains a lot of the code created before. Main differences is that the cpc doesn't change the score and that notification is made using plagiarism cases flow.

When cpc detects plagiarism, authors of all plagiarism submissions will receive a plagiarism case and they will see a _Significant Similarity_ button on the exercise participation site. This button is similar to the _Plagiarism Case_ button visible when a standard plagiarism case is created for the exercise.

The cpc is an optional feature which needs to be explicitly enabled by the instructor. It may be necessary to verify the legal compliance before enabling it.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming, Text or Modelling Exercise
- test branch with cpc triggered every 1min deployed (`plagiarism-detection-cpc-1min`)

1. Deploy Artemis locally or on ts1.
1. Enable cpc for your exercise: set the cpc checkbox and a due data in the exercise creation page.
3. Create two identical submissions by two different students.
4. Wait 1min.
5. Verify that the notification about the plagiarism case was delivered to both students and that both students and the instructor can view it (see screenshots for reference).
6. Completely change one submission.
7. Wait 1min.
8. Verify that the plagiarism case was deleted.

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlService.java | 100% | ✅                           |
| PlagiarismDetectionService.java | 100% | ✅                           |
| PlagiarismDetectionConfigHelper.java | 86% (empty constructor omitted) | ✅                           |

### Screenshots
The _Significant Similarity_ button visible on the participation page:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/a975828b-e905-43ca-a933-378d23191fc0">

The cpc plagiarism case student notification:
<img width="675" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/9d612118-b2f6-4163-8ed3-66edc81dc8a2">

The cpc plagiarism case student view:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/4ae02d20-2cfd-4738-a80d-f7bd1b43bc0e">

The cpc plagiarism case instructor view:
<img width="1324" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/5f1dfbb5-c2dd-49e0-b530-b906c37e14ed">
<img width="1324" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/d218e4ec-7261-49d4-9d81-2df1d8a969a7">

The cpc plagiarism case email:
<img width="753" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/975e0302-685e-4f6c-a09b-bf4f9f8ccf6d">
